src/ClassDiscovery.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Http\Discovery;
+
+/**
+ * Registry that based find results on class existence
+ *
+ * @author David de Boer <david@ddeboer.nl>
+ */
+abstract class ClassDiscovery
+{
+    /**
+     * Cached class
+     *
+     * @var object
+     */
+    protected static $cache;
+
+    /**
+     * Conditions and classes
+     *
+     * @var array
+     */
+    protected static $classes = [];
+
+    /**
+     * Register a factory
+     *
+     * @param string $name
+     * @param string $condition
+     * @param string $class
+     */
+    public static function register($name, $condition, $class = null)
```

Comment:
I would change the order of arguments: the first is the class, which is always required. The second is the condition. Even in factory guessers, it is acceptable to pass the condition as a second argument. Again, just for the sake of consistency.